File: app/src/main/java/com/explorify/xplore/xplore/LibraryFragment.java
package com.explorify.xplore.xplore;

import android.app.Fragment;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Nika on 11/9/2016.
 */

public class LibraryFragment extends Fragment implements TextView.OnEditorActionListener {

    private ListView list;
    private List<Integer> resultIDs = new ArrayList<>();
    private ArrayList<ReserveButton> answerButtons = new ArrayList<>();
    private ArrayList<ReserveButton> reserveButtons = new ArrayList<>();
    private DBManager dbManager;
    private View myView;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        myView = inflater.inflate(R.layout.search_layout, container, false);
        return myView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        //TODO convert this to java and skip the other crap arguments
        dbManager = new DBManager(getActivity(), "reserveDB.db", General.DB_TABLE);
        dbManager.openDataBase();

        //setting up the listview
        list = (ListView) myView.findViewById(R.id.resultslist);

        //setting up searchbar
        EditText searchBar = (EditText) myView.findViewById(R.id.search_bar);
        searchBar.setSingleLine(true);
        searchBar.setHint(R.string.search_hint);
        searchBar.setSelectAllOnFocus(true);
        searchBar.setOnEditorActionListener(this);

        //Load all reserveButtons from DB
        populateButtonList(reserveButtons);

        //Clear answer list
        answerButtons.clear();

        //Push all reserveButtons to answers
        answerButtons.addAll(reserveButtons);

        //Setting adapter
        ArrayAdapter<ReserveButton> adapter = new LibraryFragment.MyListAdapter();
        list.setAdapter(adapter);
    }

    @Override
    public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {

        searchListItems(textView.getText().toString().toLowerCase());

        return false;
    }

    private class MyListAdapter extends ArrayAdapter<ReserveButton> {
        public MyListAdapter() {
            super(getActivity(), R.layout.list_item, answerButtons);
        }

        @NonNull
        @Override
        public View getView(final int position, View convertView, ViewGroup parent) {
            View itemView = convertView;
            if (itemView == null) {
                itemView = getActivity().getLayoutInflater().inflate(R.layout.list_item, parent, false);
            }
            final ReserveButton currentButton = answerButtons.get(position);

            Button butt = (Button) itemView.findViewById(R.id.resultItem);
            butt.setText(currentButton.getName());
            butt.setBackground(currentButton.getImage());

            //Configuring Clicks
            butt.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    General.HideKeyboard(getActivity());
                    //TODO afrer changing LibFragment from FragmentActivity to Fragment, use fragmentmanager to replace fragment_containter here
                    General.OpenLibFragment(currentButton.getId(), getActivity());
                }
            });

            return itemView;
        }
    }

    //TODO after converting to kotlin, do this asynchronously
    private void populateButtonList(ArrayList<ReserveButton> reserveButtons)
    {
        String table = General.DB_TABLE;
        Resources resources = getActivity().getResources();

        reserveButtons.clear();

        //Getting each resID separately
        //TODO this is utter shit, put the loop inside of DBManager so it doesn't create and destroy a goddamn cursor every time we need a string from DB
        for(int i = 0; i < MainActivity.RESERVE_NUM; i++)
        {
            //TODO omg change this
            int resid = resources.getIdentifier(dbManager.getStr(i, DBManager.ColumnNames.getIMAGE(), table),"drawable","com.explorify.xplore.xplore");
            reserveButtons.add (
                    new ReserveButton(i, ContextCompat.getDrawable(getActivity(), resid),
                            dbManager.getStr(i, DBManager.ColumnNames.getNAME(), table))
            );
        }
    }

    //Searches DB for query
    private void searchListItems(String query) {
        answerButtons.clear();

        //Searching Database
        //TODO do getCurrentTable once on a private String for fucks sake (or do it in DBManager once on init)
        resultIDs = dbManager.getIdFromQuery(query, General.DB_TABLE);

        //Returning Results
        if(resultIDs == null)
        {
            Toast.makeText(getActivity(), R.string.search_no_results , Toast.LENGTH_SHORT).show();
        }
        else {
            int index = 0;
            for (int result : resultIDs) {
                //result is the single ID of an answer
                answerButtons.add(index,reserveButtons.get(result));
                index ++;
            }

            //Setting adapter
            ArrayAdapter<ReserveButton> adapter = new LibraryFragment.MyListAdapter();
            list.setAdapter(adapter);
        }
    }
}

File: app/src/main/java/com/explorify/xplore/xplore/LibraryFragment.kt
package com.explorify.xplore.xplore

import android.app.Activity
import android.app.Fragment
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*

import java.util.ArrayList

import kotlinx.android.synthetic.main.list_item_reserve.view.*
import kotlinx.android.synthetic.main.search_layout2.resultsRV
import kotlinx.android.synthetic.main.search_layout2.searchEditText
import kotlinx.android.synthetic.main.search_layout2.progressBar

/**
 * Created by Nika on 11/9/2016.
 */

class LibraryFragment : Fragment(), TextView.OnEditorActionListener {

    private var resultIDs: List<Int>? = ArrayList()
    private val answerCards = ArrayList<ReserveCard>()
    private val dbManager: DBManager by lazy { DBManager(activity) }
    private val reserveCards by lazy { dbManager.getAllReserveCards() }
    private var firstLoad = true



    //TODO wait for nav bar animation to close, then start loading



    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.search_layout2, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    fun init(){
        //setting up searchbar
        searchEditText.setSingleLine(true)
        searchEditText.setHint(R.string.search_hint)
        searchEditText.setSelectAllOnFocus(true)
        searchEditText.setOnEditorActionListener(this)

        //Clear answer list
        answerCards.clear()

        //Setting layoutmanager
        resultsRV.layoutManager = LinearLayoutManager(activity)

        //Starting the loading animation
        //progressBar.visibility = View.VISIBLE
    }

    fun firstDisplayData()
    {
        dbManager.openDataBase()

        //Creating adapter
        val adapter = RVadapter(reserveCards, activity)
        resultsRV.adapter = adapter

/*        //Load all reserveCards dynamically
        Thread(Runnable {
            resultsRV.post { adapter.notifyDataSetChanged() }
            progressBar.post { progressBar.visibility = View.INVISIBLE }
        }).start()*/
    }

    override fun onResume() {
        super.onResume()
        if(firstLoad){
            firstLoad = false
            firstDisplayData()
        }
    }

    class RVadapter(val results: List<ReserveCard>, val activity: Activity) : RecyclerView.Adapter<RVadapter.ResultViewHolder>(){

        class ResultViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView){
            internal val reserveName: TextView
            internal val reserveImage: ImageView
            internal val reserveIcon: ImageView

            init{
                reserveName = itemView.reserveNameTextView
                reserveImage = itemView.reserveImageView
                reserveIcon = itemView.reserveIconImageView
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ResultViewHolder =
                ResultViewHolder(
                        LayoutInflater.from(parent.context)
                                .inflate(R.layout.list_item_reserve, parent, false)
                )

        override fun onBindViewHolder(holder: ResultViewHolder, position: Int) {
            holder.reserveName.setText(results[position].name)
            holder.reserveImage.setImageResource(results[position].imageId)
            //holder.reserveImage.setImageResource(results[position].iconId)
            holder.itemView.setOnClickListener {
                General.HideKeyboard(activity)
                General.openReserveInfoFragment(results[position].id, activity)
            }
        }

        override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
            super.onAttachedToRecyclerView(recyclerView)
        }

        override fun getItemCount(): Int = results.size
    }

    //TODO change this in dbManager to just return all cards instead of the loop here
    private fun populateCardList(reserveCards: ArrayList<ReserveCard>, dbManager: DBManager) {
        val table = General.DB_TABLE
        reserveCards.clear()

        Thread(Runnable {
            //Getting each resID separately
            for (i in 0..MainActivity.RESERVE_NUM - 1) {
                reserveCards.add(dbManager.getReserveCard(i))
                resultsRV.post { resultsRV.adapter.notifyDataSetChanged() }
            }
        }).start()

    }

    override fun onEditorAction(textView: TextView, i: Int, keyEvent: KeyEvent): Boolean {
        //TODO this is a hack, change this, don't create new dbManager on each event
        //TODO lazy load dbManager
        searchListItems(textView.text.toString().toLowerCase(), dbManager)
        return false
    }

    //Searches DB for query
    private fun searchListItems(query: String, dbManager: DBManager) {
        answerCards.clear()

        //Searching Database
        resultIDs = dbManager.getIdFromQuery(query, General.DB_TABLE)

        //Returning Results
        if (resultIDs == null) {
            Toast.makeText(activity, R.string.search_no_results, Toast.LENGTH_SHORT).show()
        } else {
            var index = 0
            for (result in resultIDs!!) {
                //result is the single ID of an answer
                answerCards.add(index, reserveCards[result])
                index++
            }

            //Setting adapter
            resultsRV.adapter = RVadapter(answerCards, activity)
        }
    }
}
